### ISSUE TITLE
hsm_custom_lib > Compile issue

### ISSUE THREAD

**Issue by qzzf1w** [Azure/azure-iot-sdk-c#1643]
Am on Raspberry PI (Latest version)

Am trying to compile the custom hsm sample.

Running into an issue recognizing libcustom_hsm_example.a

[Command]---------------------------------------------------------

cmake -Duse_prov_client:BOOL=ON -Dhsm_custom_lib="~/azure-iot-sdk-c/cmake/provisioning_client/samples/custom_hsm_example/libcustom_hsm_example.a" .
cmake --build .

[Errror]---------------------------------------------------------

[ 74%] Building C object iothub_client/CMakeFiles/iothub_client.dir/src/blob.c.o
[ 75%] Linking C static library libiothub_client.a
[ 75%] Built target iothub_client
Scanning dependencies of target iothub_convenience_sample
[ 75%] Building C object iothub_client/samples/iothub_convenience_sample/CMakeFiles/iothub_convenience_sample.dir/iothub_convenience_sample.c.o
make[2]: *** No rule to make target '/root/azure-iot-sdk-c/cmake/provisioning_client/samples/custom_hsm_example/libcustom_hsm_example.a', needed by 'iothub_client/samples/iothub_convenience_sample/iothub_convenience_sample'.  Stop.
make[1]: *** [CMakeFiles/Makefile2:2201: iothub_client/samples/iothub_convenience_sample/CMakeFiles/iothub_convenience_sample.dir/all] Error 2
make: *** [Makefile:141: all] Error 2

**Comment by danewalton** [comment]
Hi @qzzf1w can you try adding this -> `-Dhsm_type_custom=ON` to your cmake command?

**Comment by qzzf1w** [comment]
And one more thing.

All of this works on windows

**Comment by danewalton** [comment]
Okay yea so unfortunately I think the CMake situation could be better here.
TLDR to get you unblocked:
```bash
cmake -Duse_prov_client=ON ..
cmake --build .
cmake -Duse_prov_client:BOOL=ON -Dhsm_type_custom=ON -Dhsm_custom_lib="~/azure-iot-sdk-c/cmake/provisioning_client/samples/custom_hsm_example/libcustom_hsm_example.a" .
cmake --build .
```

Reasoning being the first pass, it will build the custom lib and then the second pass it will link that newly built lib in your solution.

Let me know if that works. I can look to get this to work better on a first pass only but for now that's how it'll work.

**Comment by qzzf1w** [comment]
I beleive I did two pass but will try again
Stay posted

**Comment by qzzf1w** [comment]
Here is what I am using (Latest Raspbian build)

----------------------------------------------------------------------
Operating System: Raspbian GNU/Linux 10 (buster)
                  Kernel: Linux 5.4.51+
-----------------------------------------------------

My precise steps to repro as follows

1. Install new Raspbian image on Micro SD Card
2.  sudo apt-get install cmake build-essential
    sudo apt-get install git-core
    sudo apt-get purge -y libssl-dev
    sudo apt-get install libcurl4-openssl-dev
    sudo apt-get install libssl-dev
    sudo apt-get update
    sudo apt-get upgrade
    sudo reboot
3.  sudo git clone -b LTS_07_2020_Ref01 --recursive https://github.com/azure/azure-iot-sdk-c.git
    cd azure-iot-sdk-c
    sudo mkdir cmake
    cd cmake
    sudo cmake -Duse_prov_client:BOOL=ON ..
    sudo cmake --build .
   sudo cmake -Duse_prov_client:BOOL=ON -Dhsm_type_custom=ON -Dhsm_custom_lib="~/azure-iot-sdk-c/cmake/provisioning_client/samples/custom_hsm_example/libcustom_hsm_example.a" .
   sudo cmake --build .

Same results

[Error]------------------------
[ 74%] Building C object iothub_client/CMakeFiles/iothub_client.dir/src/blob.c.o
[ 75%] Linking C static library libiothub_client.a
[ 75%] Built target iothub_client
Scanning dependencies of target iothub_convenience_sample
[ 75%] Building C object iothub_client/samples/iothub_convenience_sample/CMakeFiles/iothub_convenience_sample.dir/iothub_convenience_sample.c.o
make[2]: *** No rule to make target '/root/azure-iot-sdk-c/cmake/provisioning_client/samples/custom_hsm_example/libcustom_hsm_example.a', needed by 'iothub_client/samples/iothub_convenience_sample/iothub_convenience_sample'.  Stop.
make[1]: *** [CMakeFiles/Makefile2:2201: iothub_client/samples/iothub_convenience_sample/CMakeFiles/iothub_convenience_sample.dir/all] Error 2
make: *** [Makefile:141: all] Error 2

**Comment by danewalton** [comment]
actually maybe try expanding out the relative path of the custom lib?

**Comment by qzzf1w** [comment]
That is it ~ resolves to "root" I wanted "home/pi"

**Comment by danewalton** [comment]
Nice! Alright so I'll go ahead and close this for now. Let me know if you want it reopened for any reason.

**Comment by az-iot-builder-01** [comment]
@danewalton, @qzzf1w, thank you for your contribution to our open-sourced project! Please help us improve by filling out this 2-minute [customer satisfaction survey](https://forms.office.com/Pages/ResponsePage.aspx?id=v4j5cvGGr0GRqy180BHbRx8jpG79OudPqY-BojcKjkJUMFFBQjgxNkw5ODJIRjBWTzI0Nk85R0pUQiQlQCN0PWcu)

**Comment by qzzf1w** [comment]
Great support Dane!
